sklearn/impute/_iterative.py

```diff
@@ -743,7 +766,17 @@ def fit_transform(self, X, y=None):
                     "[IterativeImputer] Early stopping criterion not reached.",
                     ConvergenceWarning,
                 )
-        Xt[~mask_missing_values] = X[~mask_missing_values]
+        for feat_idx, feat_mask in enumerate(mask_missing_values.T):
```

Comment:
Does this introduce a performance regression when everything is a `ndarray`?

Reply:
It looks like a slowdown of x1.5

Reply:
Could we add a disjunction based on whether `mask_missing_values` is a `np.array` or a `pd.DataFrame`?

Reply:
Let's write a small helper function because the same pattern also appears in `transform`.

Reply:
I made a distinction for whether Xt is a dataframe or not. Also pandas has a method for this specific use case. I put the logic in a function named `_assign_where`.